Source/DafnyCore/AST/TypeUtil.cs

```diff
@@ -0,0 +1,140 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+
+namespace Microsoft.Dafny;
+
+public class TypeUtil {
+  public static Dictionary<TypeParameter, Type> TypeSubstitutionMap(List<TypeParameter> formals, List<Type> actuals) {
```

Comment:
Consider removing `Type` from the name since that's already part of `TypeUtil`. Also consider moving this to `TypeParameter` since it strongly related to it.